src/model/differ.js

```diff
@@ -168,9 +191,23 @@ export default class Differ {
 	 * the position on which the change happened. If a position {@link module:engine/model/position~Position#isBefore is before}
 	 * another one, it will be on an earlier index in the diff set.
 	 *
+	 * Because calculating diff is a costly operation, the result is cached. If no new operation was buffered since the
+	 * previous {@link #getChanges} call, the next call with return the cached value.
+	 *
+	 * @params {Boolean} [includeChangesInGraveyard=false] If set to `true`, also changes that happened in graveyard root will be returned.
 	 * @returns {Array.<Object>} Diff between old and new model tree state.
 	 */
-	getChanges() {
+	getChanges( includeChangesInGraveyard = false ) {
+		if ( this._cachedChanges ) {
+			// If there are cached changes, just return them instead of calculating changes again.
+			if ( !includeChangesInGraveyard ) {
+				// Filter out graveyard changes.
+				return this._cachedChanges.slice().filter( _changesInGraveyardFilter );
```

Comment:
You check the root of every position in the cached changes. It's not very effective, especially if in most cases changes in the graveyard will not be used. Can't we just keep the information if the cashed changes include changes in the graveyard? Or have 2 caches (in one store change with and in the other without these in graveyard)?

Reply:
I don't think that's that ineffective. It's just one check for each `entry`. `root` is just a simple property of `Range` or `Position` (it is not evaluated). I can make two caches if you really want to but I don't think this is that crucial.

Reply:
We talked F2F, and @scofalik realized what is the problem here.